client-server-client/src/main/scala/zio/raft/client/RaftClient.scala

```diff
@@ -0,0 +1,695 @@
+   */
+  def handleServerMessage(message: ServerMessage): UIO[Unit] = 
+    for {
+      // Handle promise completion for session operations
+      _ <- message match {
+        case SessionCreated(sessionId, nonce) =>
+          completeSessionOperation(nonce, sessionId)
+        case SessionContinued(nonce) =>
+          completeSessionContinuation(nonce)
+        case rejection: SessionRejected =>
+          failSessionOperation(rejection.nonce, rejection.reason)
+        case response: ClientResponse =>
+          retryManager.completePendingRequest(response.requestId, response.result)
+        case error: RequestError =>
+          retryManager.completePendingRequestWithError(error.requestId, error.reason)
+        case _ =>
+          ZIO.unit
+      }
+      
+      // Handle state transition
+      currentState <- stateRef.get
+      newState <- currentState.handleMessage(message)
+      _ <- if (newState != currentState) {
+        stateRef.set(newState) *> ZIO.logDebug(s"State transition: ${currentState.stateName} -> ${newState.stateName}")
+      } else {
+        ZIO.unit
+      }
+    } yield ()
+  
+  private def completeSessionOperation(nonce: Nonce, sessionId: SessionId): UIO[Unit] =
+    for {
+      operationsMap <- pendingOperationsRef.get
+      _ <- operationsMap.get(nonce) match {
+        case Some(SessionOperation.Create(promise)) =>
+          promise.succeed(sessionId).ignore *> pendingOperationsRef.update(_.removed(nonce))
+        case _ =>
+          ZIO.logWarning(s"Received SessionCreated for unexpected nonce: $nonce")
+      }
+    } yield ()
+  
+  private def completeSessionContinuation(nonce: Nonce): UIO[Unit] =
+    for {
+      operationsMap <- pendingOperationsRef.get
+      _ <- operationsMap.get(nonce) match {
+        case Some(SessionOperation.Continue(promise)) =>
+          promise.succeed(()).ignore *> pendingOperationsRef.update(_.removed(nonce))
+        case _ =>
+          ZIO.logWarning(s"Received SessionContinued for unexpected nonce: $nonce")
+      }
+    } yield ()
+  
+  private def failSessionOperation(nonce: Nonce, reason: RejectionReason): UIO[Unit] =
+    for {
+      operationsMap <- pendingOperationsRef.get
+      _ <- operationsMap.get(nonce) match {
+        case Some(SessionOperation.Create(promise)) =>
+          val error = SessionError.fromRejectionReason(reason)
+          promise.fail(error).ignore *> pendingOperationsRef.update(_.removed(nonce))
+        case Some(SessionOperation.Continue(promise)) =>
+          val error = SessionError.fromRejectionReason(reason)
+          promise.fail(error).ignore *> pendingOperationsRef.update(_.removed(nonce))
+        case _ =>
+          ZIO.logWarning(s"Received SessionRejected for unexpected nonce: $nonce")
+      }
+    } yield ()
+  
+  /**
+   * Get current connection state.
+   */
+  def getState: UIO[ClientState] = 
+    stateRef.get
+  
+  /**
+   * Check if client is connected.
+   */
+  def isConnected: UIO[Boolean] = 
```

Comment:
we probably don't need this, if we really want, we should enqueue an action with a promise, and the stream will handle it.